zend_abstract_interface/interceptor/php7/interceptor.c

```diff
@@ -0,0 +1,580 @@
+    prev_return_handler = zend_get_user_opcode_handler(ZEND_RETURN);
+    user_opcode_handler_t return_handler = prev_return_handler ? zai_interceptor_return_handler : zai_interceptor_return_handler_no_prev;
+    zend_set_user_opcode_handler(ZEND_RETURN, return_handler);
+    prev_return_by_ref_handler = zend_get_user_opcode_handler(ZEND_RETURN_BY_REF);
+    zend_set_user_opcode_handler(ZEND_RETURN_BY_REF, zai_interceptor_return_by_ref_handler);
+    prev_generator_return_handler = zend_get_user_opcode_handler(ZEND_GENERATOR_RETURN);
+    zend_set_user_opcode_handler(ZEND_GENERATOR_RETURN, zai_interceptor_generator_return_handler);
+    prev_handle_exception_handler = zend_get_user_opcode_handler(ZEND_HANDLE_EXCEPTION);
+    zend_set_user_opcode_handler(ZEND_HANDLE_EXCEPTION, zai_interceptor_handle_exception_handler);
+    prev_fast_ret_handler = zend_get_user_opcode_handler(ZEND_FAST_RET);
+    zend_set_user_opcode_handler(ZEND_FAST_RET, zai_interceptor_fast_ret_handler);
+
+#ifndef ZTS
+    ZEND_VM_SET_OPCODE_HANDLER(EG(exception_op));
+    ZEND_VM_SET_OPCODE_HANDLER(EG(exception_op)+1);
+    ZEND_VM_SET_OPCODE_HANDLER(EG(exception_op)+2);
+#endif
+
+    // generator intercepting weirdness
+    generator_create_prev = zend_ce_generator->create_object;
+    zend_ce_generator->create_object = zai_interceptor_generator_create;
+#if PHP_VERSION_ID >= 70100
+#define ZAI_INTERCEPTOR_POST_GENERATOR_CREATE_OP 224 // random 8 bit number greater than ZEND_VM_LAST_OPCODE
+    prev_post_generator_create_handler = zend_get_user_opcode_handler(ZAI_INTERCEPTOR_POST_GENERATOR_CREATE_OP);
+    zend_set_user_opcode_handler(ZAI_INTERCEPTOR_POST_GENERATOR_CREATE_OP, zai_interceptor_post_generator_create_handler);
+    prev_generator_create_handler = zend_get_user_opcode_handler(ZEND_GENERATOR_CREATE);
+    zend_set_user_opcode_handler(ZEND_GENERATOR_CREATE, zai_interceptor_generator_create_handler);
+
+    zai_interceptor_generator_create_wrapper[0].opcode = ZAI_INTERCEPTOR_POST_GENERATOR_CREATE_OP;
+    SET_UNUSED(zai_interceptor_generator_create_wrapper[0].result);
+    SET_UNUSED(zai_interceptor_generator_create_wrapper[0].op1);
+    SET_UNUSED(zai_interceptor_generator_create_wrapper[0].op2);
+    ZEND_VM_SET_OPCODE_HANDLER(&zai_interceptor_generator_create_wrapper[0]);
+    zai_interceptor_generator_create_wrapper[1].opcode = ZAI_INTERCEPTOR_POST_GENERATOR_CREATE_OP;
+    SET_UNUSED(zai_interceptor_generator_create_wrapper[1].result);
+    SET_UNUSED(zai_interceptor_generator_create_wrapper[1].op1);
+    SET_UNUSED(zai_interceptor_generator_create_wrapper[1].op2);
+    ZEND_VM_SET_OPCODE_HANDLER(&zai_interceptor_generator_create_wrapper[1]);
+#endif
+
+    INIT_NS_CLASS_ENTRY(zai_interceptor_bailout_ce, "Zend Abstract Interface", "BailoutHandler", NULL);
+    zai_interceptor_bailout_ce.type = ZEND_INTERNAL_CLASS;
+    zend_initialize_class_data(&zai_interceptor_bailout_ce, false);
+    zai_interceptor_bailout_ce.info.internal.module = module_entry;
```

Comment:
This doesn't need to be an actual module, it can be a dummy, right ?

Reply:
Another approach might be to add a dummy extension api to tea, we can export a module structure, I just don't want to export the one we use internally in tea ...

Reply:
yes, for the test it might be dummy memory, but in real execution at least something like xdebug will access these.